haystack_experimental/evaluation/harness/rag/harness.py

```diff
@@ -0,0 +1,355 @@
+
+from ...util.helpers import (
+    aggregate_batched_pipeline_outputs,
+    deaggregate_batched_pipeline_inputs,
+)
+from ...util.pipeline_pair import PipelinePair
+from ..evalution_harness import EvaluationHarness
+from .evaluation_pipeline import default_rag_evaluation_pipeline
+from .parameters import (
+    RAGEvaluationInput,
+    RAGEvaluationMetric,
+    RAGEvaluationOutput,
+    RAGEvaluationOverrides,
+    RAGExpectedComponent,
+    RAGExpectedComponentMetadata,
+)
+
+
+class RAGEvaluationHarness(EvaluationHarness[RAGEvaluationInput, RAGEvaluationOverrides, RAGEvaluationOutput]):
+    """
+    Evaluation harness for evaluating RAG pipelines.
+    """
+
+    def __init__(
+        self,
+        rag_pipeline: Pipeline,
+        rag_components: Dict[RAGExpectedComponent, RAGExpectedComponentMetadata],
+        metrics: Set[RAGEvaluationMetric],
+    ):
+        """
+        Create a evaluation harness for evaluating basic RAG pipelines.
+
+        :param rag_pipeline:
+            The RAG pipeline to evaluate.
+        :param rag_components:
+            A mapping of expected components to their metadata.
+        :param metrics:
+            The metrics to use during evaluation.
+        """
+        super().__init__()
+
+        self._validate_rag_components(rag_pipeline, rag_components)
+
+        self.rag_pipeline = rag_pipeline
+        self.rag_components = rag_components
+        self.metrics = metrics
+        self.evaluation_pipeline = default_rag_evaluation_pipeline(metrics)
+
+    @classmethod
+    def default_with_embedding_retriever(
+        cls, rag_pipeline: Pipeline, metrics: Set[RAGEvaluationMetric]
+    ) -> "RAGEvaluationHarness":
+        """
+        Create a default evaluation harness for evaluating RAG pipelines with a query embedder.
+
+        :param rag_pipeline:
+            The RAG pipeline to evaluate. The following assumptions are made:
+            - The query embedder component is named 'query_embedder' and has a 'text' input.
+            - The document retriever component is named 'retriever' and has a 'documents' output.
+            - The response generator component is named 'generator' and has a 'replies' output.
+        :param metrics:
+            The metrics to use during evaluation.
+        """
+        rag_components = {
+            RAGExpectedComponent.QUERY_PROCESSOR: RAGExpectedComponentMetadata(
+                name="query_embedder", input_mapping={"query": "text"}
+            ),
+            RAGExpectedComponent.DOCUMENT_RETRIEVER: RAGExpectedComponentMetadata(
+                name="retriever", output_mapping={"retrieved_documents": "documents"}
+            ),
+            RAGExpectedComponent.RESPONSE_GENERATOR: RAGExpectedComponentMetadata(
+                name="generator", output_mapping={"replies": "replies"}
+            ),
+        }
+
+        return cls(rag_pipeline, rag_components, deepcopy(metrics))
+
+    @classmethod
+    def default_with_keyword_retriever(
```

Comment:
The motivation behind having separate "default" functions is to make it as simple as possible for the user to initialize a pipeline without having to understand the mapping of the expected components. If we were to merge the two into a single function, we'd have to expose some kind of enumeration to select the type of query processor component (as it can't be a simple boolean flag given the semantics of the parameter), which introduces more code than it removes.